File: js/parser.js
function parseSJDF(input) {
  const lines = input.split("\n");
  const stack = [{}];
  const indentStack = [0];

  for (let rawLine of lines) {
    let line = rawLine.trim();
    if (!line || line.startsWith("#")) continue;

    let indent = rawLine.search(/\S/);

    while (indent < indentStack[indentStack.length - 1]) {
      stack.pop();
      indentStack.pop();
    }

    let current = stack[stack.length - 1];

    if (line.startsWith("- ")) {
      let value = parseValue(line.slice(2));
      if (!Array.isArray(current._array)) {
        current._array = [];
      }
      current._array.push(value);
      continue;
    }

    let [key, ...rest] = line.split(":");
    let value = rest.join(":").trim();

    if (!value) {
      current[key] = {};
      stack.push(current[key]);
      indentStack.push(indent);
    } else {
      current[key] = parseValue(value);
    }
  }

  return cleanArrays(stack[0]);
}

function parseValue(value) {
  if (value === "true") return true;
  if (value === "false") return false;
  if (!isNaN(value)) return Number(value);

  if (value.startsWith("[") && value.endsWith("]")) {
    return value
      .slice(1, -1)
      .split(",")
      .map(v => parseValue(v.trim()));
  }

  return value.replace(/^"|"$/g, "");
}

function cleanArrays(obj) {
  if (obj._array) return obj._array;

  for (let key in obj) {
    obj[key] = cleanArrays(obj[key]);
  }

  return obj;
}

module.exports = parseSJDF;

File: python/parser.py
def parse_sjdf(text):
    lines = text.split("\n")
    stack = [{}]
    indent_stack = [0]

    for raw_line in lines:
        if not raw_line.strip() or raw_line.strip().startswith("#"):
            continue

        indent = len(raw_line) - len(raw_line.lstrip())
        line = raw_line.strip()

        while indent < indent_stack[-1]:
            stack.pop()
            indent_stack.pop()

        current = stack[-1]

        if line.startswith("- "):
            value = parse_value(line[2:])
            current.setdefault("_array", []).append(value)
            continue

        parts = line.split(":")
        key = parts[0]
        value = ":".join(parts[1:]).strip()

        if not value:
            current[key] = {}
            stack.append(current[key])
            indent_stack.append(indent)
        else:
            current[key] = parse_value(value)

    return clean_arrays(stack[0])


def parse_value(value):
    if value == "true":
        return True
    if value == "false":
        return False

    try:
        return int(value)
    except:
        pass

    if value.startswith("[") and value.endswith("]"):
        return [parse_value(v.strip()) for v in value[1:-1].split(",")]

    return value.strip('"')


def clean_arrays(obj):
    if isinstance(obj, dict):
        if "_array" in obj:
            return obj["_array"]
        return {k: clean_arrays(v) for k, v in obj.items()}
    return obj
